exchanges/kline/request.go

```diff
@@ -0,0 +1,185 @@
+	// Pair refers to the currency pair
+	Pair currency.Pair
+	// RequestFormatted refers to the currency pair formatted by the exchange
+	// asset for outbound requests
+	RequestFormatted currency.Pair
+	// Asset refers to the asset type
+	Asset asset.Item
+	// ExchangeInterval refers to the interval that is used to construct the
+	// client required interval, this will be less than or equal to the client
+	// required interval.
+	ExchangeInterval Interval
+	// ClientRequired refers to the clients' actual required interval
+	// needed.
+	ClientRequired Interval
+	// Start is the start time aligned to UTC and to the Required interval candle
+	Start time.Time
+	// End is the end time aligned to UTC and to the Required interval candle
+	End time.Time
+}
+
+// CreateKlineRequest generates a `Request` type for interval conversions
+// supported by an exchange.
+func CreateKlineRequest(name string, pair, formatted currency.Pair, a asset.Item, clientRequired, exchangeInterval Interval, start, end time.Time) (*Request, error) {
+	if name == "" {
+		return nil, ErrUnsetName
+	}
+	if pair.IsEmpty() {
+		return nil, currency.ErrCurrencyPairEmpty
+	}
+	if formatted.IsEmpty() {
+		return nil, currency.ErrCurrencyPairEmpty
+	}
+	if !a.IsValid() {
+		return nil, asset.ErrNotSupported
+	}
+	if clientRequired == 0 {
+		return nil, fmt.Errorf("client required %w", ErrInvalidInterval)
+	}
+	if exchangeInterval == 0 {
+		return nil, fmt.Errorf("exchange interval %w", ErrInvalidInterval)
+	}
+	err := common.StartEndTimeCheck(start, end)
+	if err != nil {
+		return nil, err
+	}
+
+	// Force UTC alignment
+	start = start.UTC()
+	end = end.UTC()
+
+	// Force alignment to required interval which is the higher time value e.g.
+	// 1hr required as opposed to 1min request/outbound interval used to
+	// construct the higher time value candle. This is to make sure there are
+	// minimal missing candles which is used to create the bigger candle.
+	start = start.Truncate(clientRequired.Duration())
+
+	// Strip monotonic clock reading for comparison
+	end = end.Round(0)
+
+	endTrunc := end.Truncate(clientRequired.Duration())
+	// Check to see if truncation moves end time and if so we want to make sure
+	// the candle period is included on the end.
+	forward := endTrunc.Add(clientRequired.Duration())
+	if !endTrunc.Equal(end) && !forward.After(time.Now()) {
+		end = forward
+	}
+	return &Request{name, pair, formatted, a, exchangeInterval, clientRequired, start, end}, nil
+}
+
+// GetRanges returns the date ranges for candle intervals broken up over
+// requests
+func (r *Request) GetRanges(limit uint32) (*IntervalRangeHolder, error) {
+	if r == nil {
+		return nil, errNilRequest
+	}
+	return CalculateCandleDateRanges(r.Start, r.End, r.ExchangeInterval, limit)
+}
+
+// ConvertCandles converts time series candles into a kline.Item type. This will
+// auto convert from a lower to higher time series if applicable.
+func (r *Request) ConvertCandles(timeSeries []Candle) (*Item, error) {
```

Comment:
I'm going to suggest that the majority of requests are not going to be fancy custom intervals, which means that converting candles does not happen much and so this name is a bit off. `ProcessResponse` is vague too. I'm not getting too upset about this, but calling it `ConvertCandles` may lead to other devs not using this function because they do not wish to change the candle to something else which could cause some issues since so much additional processing occurs in here - which was why my suggestion was `ProcessResponse`